### PR TITLE
Fix special tokens not correctly tokenized

### LINKED ISSUE #13483: `BertTokenizer` splits tokens added via `add_tokens(...', special_tokens=False)` to a few subtokens
## Environment

`!transformers-cli env` output:

- `transformers` version: 4.11.0.dev0
- Platform: Linux-5.4.104+-x86_64-with-Ubuntu-18.04-bionic
- Python version: 3.7.11
- PyTorch version (GPU?): 1.9.0+cu102 (False)
- Tensorflow version (GPU?): 2.6.0 (False)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using GPU in script?: no
- Using distributed or parallel set-up in script?: no

## Information

I'm using `BertTokenizer` to process data and train my BERT-based model.

I found that when I add a few "technical" tokens via `tokenizer,add_tokens(token_text, special_tokens=True)` - it still splits added tokens the same way ad before adding them into tokenizer. When I use `tokenizer,add_tokens(token_text, special_tokens=False)` instead - it works properly.

Moreover, when I save the tokenizer with `special_tokens=True`-adding through `tokenizer.save_pretrained(directory_name)` and then load it back - it works properly, instead of just modifying the tokenizer.

There I'll give a sample (you can also see the same sample here - https://colab.research.google.com/drive/1Ufa79vrvdoyee6xiz9WCR7F2UKZuBfS7?usp=sharing ):

```python
from transformers import BertTokenizer


tokenizer_base = BertTokenizer.from_pretrained("bert-base-uncased")
tokenizer_special = BertTokenizer.from_pretrained("bert-base-uncased")
tokenizer_nonspecial = BertTokenizer.from_pretrained("bert-base-uncased")

tokenizer_special.add_tokens('[TITLE]', special_tokens=True)
tokenizer_nonspecial.add_tokens('[TITLE]', special_tokens=False)

# [101, 1031, 2516, 1033, 102]
# so basically, before any modification - it split `[TITLE]` to `1031, 2516, 1033`
print(tokenizer_base('[TITLE]')['input_ids'])

# [101, 1031, 2516, 1033, 102]
# Guess it should give 3 token ids here - yet it gives same ids like `tokenizer_base`
print(tokenizer_special('[TITLE]')['input_ids'])

# [101, 30522, 102]
# So with `.add_tokens('[TITLE]', special_tokens=False)` it maps `[TITLE]` to one token
print(tokenizer_nonspecial('[TITLE]')['input_ids'])

# Guess it should give same result as `tokenizer_special`'s one
# But it gives me `[101, 30522, 102]` instead - like `tokenizer_nonspecial`'s one
tokenizer_special.save_pretrained("pretrained-with-special-flag")
tokenizer_special_from_pretrained = BertTokenizer.from_pretrained("pretrained-with-special-flag")
tokenizer_special_from_pretrained('[TITLE]')['input_ids']
```

## Expected behavior
I expected to see 1 token id for `tokenizer_special` (something like `tokenizer_nonspecial`'s `[101, 30522, 102]`) - or, at least, see that `tokenizer_special_from_pretrained` reproduces `tokenizer_special`'s behavior.

### PR DESCRIPTION
**Pull request by qqaatw**
# What does this PR do?

Fixes #13483.

I think this is a quite complex issue. The purpose of `unique_no_split_tokens` in `tokenization_utils.py` and `all_special_tokens` in `tokenization_utils_base.py` seems ambiguous as they all store special tokens that don't want to be splitted or lower_cased. This two separated variables may result in a hard maintainable code. I'm guessing this is probably due to the workaround for FastTokenizers' functionality.

Additionally, should we resolve this TODO?

https://github.com/huggingface/transformers/blob/c37573806ab3526dd805c49cbe2489ad4d68a9d7/src/transformers/tokenization_utils.py#L275-L282

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@sgugger @LysandreJik

### CODE REVIEW

**Review by sgugger — APPROVED** [review]
LGTM, thanks for fixing!

**Review by LysandreJik — COMMENTED** [review]
This is a great catch @qqaatw, thank you for opening a PR! I agree with your fix (cc @SaulLu for visibility)

Do you mind adding a test that would have failed before your fix, and that works thanks to your fix? Let me know if you need any pointers, the tests are in the `tests/test_tokenization_common.py` file!

**Comment by qqaatw** [comment]
A test has been added. Thanks for the review!

File: tests/test_tokenization_common.py [hunk 1]
```diff
@@ -310,6 +310,31 @@ def convert_batch_encode_plus_format_to_encode_plus(batch_encode_plus_sequences)
...
+                SPECIAL_TOKEN_1 = "[SPECIAL_TOKEN_1]"
+                SPECIAL_TOKEN_2 = "[SPECIAL_TOKEN_2]"
+
+                # TODO: Can we replace `unique_no_split_tokens` and `all_special_tokens` with one variable(property) for a better maintainability?
+                # The list in which `_add_tokens` method stores special tokens. (in tokenization_utils.py)
+                tokenizer.unique_no_split_tokens = [SPECIAL_TOKEN_1]
+                # The property storing additional special tokens which occur in `all_special_tokens`. (in tokenization_utils_base.py)
+                tokenizer.additional_special_tokens = [SPECIAL_TOKEN_2]
+                # Add all special tokens into `unique_no_split_tokens`.
+                tokenizer.sanitize_special_tokens()
```
Comment (LysandreJik) [inline tests/test_tokenization_common.py:328]: How about leveraging the `tokenizer.add_special_tokens` method so that this follows a user workflow?

Comment (qqaatw) [reply tests/test_tokenization_common.py:328]: I got your point. However, `tokenizer.add_special_tokens` method is only appropriate to test tokens expected to exist in `additional_special_token`. The method #13483 used is directly calling `tokenizer.add_tokens(token, special_tokens=True)` to add new special tokens, and the tokens added through this method will never occur in `tokenizer.all_special_tokens`. That's why we need this fix and the test.

TL;DR, I think we can use `tokenizer.add_special_tokens` to test whether special tokens stored in `tokenizer.all_special_tokens` can properly be tokenized, and, use `tokenizer.add_tokens` to test whether special tokens stored in `tokenizer.unique_no_split_tokens` can properly be tokenized.

What do you think?

Comment (LysandreJik) [reply tests/test_tokenization_common.py:328]: Yes that makes sense

**Review by LysandreJik — APPROVED** [review]
Looks good to me! Thank you for working on it @qqaatw
